### PR TITLE
ENG-10404 support for start and init cli commands

### PR DESCRIPTION
**Pull request by vtkstef**

### CODE REVIEW

File: lib/python/voltcli/verbs.py [hunk 1]
```diff
@@ -415,6 +416,9 @@ def __init__(self, subcommand,
         self.check_environment_config = check_environment_config
         self.force_voltdb_create = force_voltdb_create
         self.supports_paused = supports_paused
+        # this flag indicates whether or not the command is a
+        # legacy command: create, recover, rejoin, join
```
Comment (jhugg) [inline lib/python/voltcli/verbs.py]: ...as opposed to the new commands, init and start

File: src/frontend/org/voltdb/utils/MiscUtils.java [hunk 2]
```diff
@@ -526,7 +527,12 @@ public static String formatHostMetadataFromJSON(String json) {
     // check if we're running pro code
     public static boolean isPro() {
         if (m_isPro == null) {
-            m_isPro = null != MiscUtils.loadProClass("org.voltdb.CommandLogImpl", "Command logging", true);
+            //Allow running pro kit as community.
+            if (!Boolean.parseBoolean(System.getProperty("community", "false"))) {
```
Comment (jhugg) [inline src/frontend/org/voltdb/utils/MiscUtils.java]: Is this logic backwards? how about if property community = true, then m_isPro = false, otherwise do the check?

I think it serves the purpose now, but is kinda confusing.

File: src/frontend/org/voltdb/probe/MeshProber.java [hunk 3]
```diff
@@ -0,0 +1,866 @@
...
+import com.google_voltpatches.common.base.Throwables;
+import com.google_voltpatches.common.collect.ImmutableMap;
+import com.google_voltpatches.common.collect.ImmutableSortedSet;
+import com.google_voltpatches.common.collect.Maps;
+import com.google_voltpatches.common.net.HostAndPort;
+import com.google_voltpatches.common.net.InetAddresses;
+import com.google_voltpatches.common.net.InternetDomainName;
+import com.google_voltpatches.common.util.concurrent.SettableFuture;
+
+public class MeshProber implements JoinAcceptor {
```
Comment (jhugg) [inline src/frontend/org/voltdb/probe/MeshProber.java]: Top level classes really need comments explaining what they do, where they fit, who uses them, why, etc...

File: tests/frontend/org/voltdb/regressionsuites/InternalPortGeneratorForTest.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import static com.google_voltpatches.common.base.Preconditions.checkArgument;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.NavigableSet;
+
+import com.google_voltpatches.common.collect.ImmutableList;
+import com.google_voltpatches.common.collect.ImmutableSortedSet;
+
+public class InternalPortGeneratorForTest implements Iterable<Integer>{
```
Comment (jhugg) [inline tests/frontend/org/voltdb/regressionsuites/InternalPortGeneratorForTest.java]: We already have a class called PortGenerator, which is primarily for tests. Why can't this code be added there?

File: src/frontend/org/voltdb/utils/CatalogUtil.java [hunk 5]
```diff
@@ -986,6 +994,35 @@ public static String getDeployment(DeploymentType deployment) throws IOException
...
 
     /**
+     * Computes a MD5 digest (128 bits -> 2 longs -> UUID which is comprised of
+     * two longs) of a deployment file stripped of all comments and its hostcount
+     * attribute set to 0 ,and adminstartup set to false
+     *
+     * @param deploymentBytes
+     * @return MD5 digest for for configuration
+     */
+    public static UUID makeDeploymentHashForConfig(byte[] deploymentBytes) {
```
Comment (jhugg) [inline src/frontend/org/voltdb/utils/CatalogUtil.java]: Don't we already have code that does something like this somewhere? Did it get removed?

File: src/frontend/org/voltdb/utils/CatalogUtil.java [hunk 6]
```diff
@@ -986,6 +994,35 @@ public static String getDeployment(DeploymentType deployment) throws IOException
...
+        return Digester.md5AsUUID(normalized);
+    }
+
+    /**
+     * Given the deployment object generate the XML
+     * @param deployment
+     * @return XML of deployment object.
+     * @throws IOException
+     */
+    public static String getDeployment(DeploymentType deployment) throws IOException {
```
Comment (jhugg) [inline src/frontend/org/voltdb/utils/CatalogUtil.java]: Can we put this method next to its polymorph?

File: src/frontend/org/voltdb/utils/HTTPAdminListener.java [hunk 7]
```diff
@@ -100,6 +104,24 @@
     final boolean m_mustListen;
     final DeploymentRequestHandler m_deploymentHandler;
 
+    // ObjectMapper is thread safe, and uses a lot of memory to cache
```
Comment (jhugg) [inline src/frontend/org/voltdb/utils/HTTPAdminListener.java]: Is this change related to this pull request broadly?

File: src/frontend/org/voltdb/StatusTracker.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+ */
+package org.voltdb;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.voltdb.common.NodeState;
+
+import com.google_voltpatches.common.base.Supplier;
+
+public class StatusTracker {
```
Comment (jhugg) [inline src/frontend/org/voltdb/StatusTracker.java]: Please comment top level classes. And please make the comment useful. What, how, why, who, etc...

File: src/frontend/org/voltdb/RealVoltDB.java [hunk 9]
```diff
@@ -1464,6 +1522,139 @@ private void startResourceUsageMonitor() {
         }
     }
 
+    private void stageDeploymemtFileForInitialize(Configuration config, DeploymentType dt) {
```
Comment (jhugg) [inline src/frontend/org/voltdb/RealVoltDB.java]: Needs comment. WTF does this do? Why and for whom?

File: src/frontend/org/voltcore/logging/VoltLog4jLogger.java [hunk 10]
```diff
@@ -17,15 +17,23 @@
...
 import java.util.ResourceBundle;
 
+import org.apache.log4j.Appender;
+import org.apache.log4j.DailyRollingFileAppender;
 import org.apache.log4j.LogManager;
 import org.apache.log4j.Logger;
 import org.apache.log4j.xml.DOMConfigurator;
 import org.voltcore.logging.VoltLogger.CoreVoltLogger;
 import org.voltcore.utils.ShutdownHooks;
+import org.voltdb.utils.VoltFile;
```
Comment (jhugg) [inline src/frontend/org/voltcore/logging/VoltLog4jLogger.java]: voltdb leaking into voltcore.

File: src/frontend/org/voltcore/messaging/HostMessenger.java [hunk 11]
```diff
@@ -17,6 +17,12 @@
 
 package org.voltcore.messaging;
 
+import static com.google_voltpatches.common.base.Preconditions.checkArgument;
+import static com.google_voltpatches.common.base.Predicates.equalTo;
+import static com.google_voltpatches.common.base.Predicates.not;
+import static org.voltdb.VoltDB.DEFAULT_INTERNAL_PORT;
+import static org.voltdb.VoltDB.DEFAULT_ZK_PORT;
```
Comment (jhugg) [inline src/frontend/org/voltcore/messaging/HostMessenger.java]: voltdb leaking into volt core.

Maybe these posts move to volt core constants?

File: src/frontend/org/voltcore/messaging/HostMessenger.java [hunk 12]
```diff
@@ -60,10 +67,14 @@
 import org.voltcore.utils.ShutdownHooks;
 import org.voltcore.zk.CoreZK;
 import org.voltcore.zk.ZKUtil;
-import org.voltdb.VoltDB;
+import org.voltdb.StartAction;
```
Comment (jhugg) [inline src/frontend/org/voltcore/messaging/HostMessenger.java]: voltdb leaking into voltcore.

File: src/frontend/org/voltcore/messaging/JoinAcceptor.java [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+
+import java.util.Map;
+import java.util.NavigableSet;
+import java.util.Optional;
+import java.util.Set;
+
+import org.apache.zookeeper_voltpatches.ZooKeeper;
+import org.json_voltpatches.JSONObject;
+import org.json_voltpatches.JSONString;
+import org.voltdb.VersionChecker;
```
Comment (jhugg) [inline src/frontend/org/voltcore/messaging/JoinAcceptor.java]: voltdb leaking into volt core.

File: src/frontend/org/voltcore/messaging/SocketJoiner.java [hunk 14]
```diff
@@ -48,10 +51,12 @@
 import org.voltcore.logging.VoltLogger;
 import org.voltcore.network.ReverseDNSCache;
 import org.voltcore.utils.CoreUtils;
-import org.voltdb.VoltDB;
-import org.voltdb.common.Constants;
+import org.voltdb.VersionChecker;
```
Comment (jhugg) [inline src/frontend/org/voltcore/messaging/SocketJoiner.java]: voltdb leaking into voltcore.

File: src/frontend/org/voltdb/probe/HostCriteria.java [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+import org.json_voltpatches.JSONException;
+import org.json_voltpatches.JSONObject;
+import org.voltdb.StartAction;
+import org.voltdb.common.NodeState;
+
+import com.google_voltpatches.common.base.Throwables;
+import com.google_voltpatches.common.collect.ImmutableList;
+
+/**
+ *  It is a DTO that snapshots a subset of {@link MeshProber} fields, and is
```
Comment (jhugg) [inline src/frontend/org/voltdb/probe/HostCriteria.java]: What's a DTO?

File: src/frontend/org/voltdb/probe/HostCriteria.java [hunk 16]
```diff
@@ -0,0 +1,279 @@
...
+        }
+        return jo;
+    }
+
+    public boolean isUndefined() {
+        return UNDEFINED.equals(m_configHash) && UNDEFINED.equals(m_meshHash);
+    }
+
+    @Override
+    public int hashCode() {
```
Comment (jhugg) [inline src/frontend/org/voltdb/probe/HostCriteria.java]: Is this used anywhere? (comment from previous pull). And if so, can it not be?

File: src/frontend/org/voltdb/probe/MeshProber.java [hunk 17]
```diff
@@ -0,0 +1,866 @@
...
+            org.voltdb.VoltDB.crashLocalVoltDB(
+                    "interrupted while waiting to determine the cluster start action",
+                    false, e);
+        }
+        return new Determination(null,-1, false);
+    }
+
+    @Override
+    public int hashCode() {
+        final int prime = 31;
```
Comment (jhugg) [inline src/frontend/org/voltdb/probe/MeshProber.java]: Is this hashCode used anywhere? (comment from previous pull). And if so, can it not be?

File: src/frontend/org/voltdb/probe/MeshProber.java [hunk 18]
```diff
@@ -0,0 +1,866 @@
...
+
+        private Determination(StartAction startAction, int hostCount, boolean paused) {
+            this.startAction = startAction;
+            this.hostCount = hostCount;
+            this.paused = paused;
+        }
+
+        @Override
+        public int hashCode() {
+            final int prime = 31;
```
Comment (jhugg) [inline src/frontend/org/voltdb/probe/MeshProber.java]: Same comment as with the others. Can we not implement this?

File: src/frontend/org/voltdb/probe/HostCriteria.java [hunk 19]
```diff
@@ -0,0 +1,279 @@
...
+    /**
+     * It checks whether the given {@link JSONObject} contains the
+     * fields necessary to create an instance of {@link HostCriteria}.
+     * The get[type] variant of {@link JSONObject} throws a {@link JSONException}
+     * when it does not parse to the expected types
+     */
+    public static boolean hasCriteria(JSONObject jo) {
+        try {
+            return jo != null
+                && jo.has(IS_PAUSED) && (jo.getBoolean(IS_PAUSED) ? true : true)
```
Comment (jhugg) [inline src/frontend/org/voltdb/probe/HostCriteria.java]: Comment from previous pull. The long && chain make make debugging hard. Why this over a bunch of ifs?

File: src/frontend/org/voltcore/messaging/HostMessenger.java [hunk 20]
```diff
@@ -176,6 +211,7 @@ public String toString() {
                 js.key("deadhosttimeout").value(deadHostTimeout);
                 js.key("backwardstimeforgivenesswindow").value(backwardsTimeForgivenessWindow);
                 js.key("networkThreads").value(networkThreads);
+                js.key("acceptor").value(acceptor);
```
Comment (jhugg) [inline src/frontend/org/voltcore/messaging/HostMessenger.java]: Can we make more of these keys constants?

File: src/frontend/org/voltdb/probe/HostCriteria.java [hunk 21]
```diff
@@ -0,0 +1,279 @@
...
+        if (!m_configHash.equals(o.m_configHash)) {
+            ilb.add("Mismatched deployment configuration given at database startup");
+        }
+        return ilb.build();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj)
+            return true;
```
Comment (jhugg) [inline src/frontend/org/voltdb/probe/HostCriteria.java]: Does not follow code guidelines for if.

File: src/frontend/org/voltdb/probe/MeshProber.java [hunk 22]
```diff
@@ -0,0 +1,866 @@
...
+        result = prime * result
+                + ((m_meshHash == null) ? 0 : m_meshHash.hashCode());
+        result = prime * result
+                + ((m_startAction == null) ? 0 : m_startAction.hashCode());
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj)
```
Comment (jhugg) [inline src/frontend/org/voltdb/probe/MeshProber.java]: Does not follow code guidelines for if

File: src/frontend/org/voltdb/probe/MeshProber.java [hunk 23]
```diff
@@ -0,0 +1,866 @@
...
+            int result = 1;
+            result = prime * result + hostCount;
+            result = prime * result + (paused ? 1231 : 1237);
+            result = prime * result
+                    + ((startAction == null) ? 0 : startAction.hashCode());
+            return result;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
```
Comment (jhugg) [inline src/frontend/org/voltdb/probe/MeshProber.java]: Does not follow code guidelines for if.

File: src/frontend/org/voltdb/utils/CatalogUtil.java (hunk 5 above)
Comment (vtkstef) [inline src/frontend/org/voltdb/utils/CatalogUtil.java]: no this is new. as this computes an easier to deal hash that is independent of host count and admin flag values

File: src/frontend/org/voltdb/utils/HTTPAdminListener.java (hunk 7 above)
Comment (vtkstef) [inline src/frontend/org/voltdb/utils/HTTPAdminListener.java]: it is an improvement from the initial HTTP implementation of MeshProber

File: src/frontend/org/voltdb/probe/MeshProber.java [hunk 24]
```diff
@@ -0,0 +1,866 @@
...
+        do {
+            expect = m_hostCriteria.get();
+            update = ImmutableMap.<Integer, HostCriteria>builder()
+                .putAll(Maps.filterKeys(expect, not(equalTo(hostId))))
+                .build();
+        } while (!m_hostCriteria.compareAndSet(expect, update));
+    }
+
+    @Override
+    public void detract(Set<Integer> hostIds) {
```
Comment (jhugg) [inline src/frontend/org/voltdb/probe/MeshProber.java]: Not sure "detract" is the right name. What's wrong with remove?

File: src/frontend/org/voltdb/probe/MeshProber.java [hunk 25]
```diff
@@ -0,0 +1,866 @@
...
+        do {
+            expect = m_hostCriteria.get();
+            update = ImmutableMap.<Integer, HostCriteria>builder()
+                .putAll(Maps.filterKeys(expect, not(in(hostIds))))
+                .build();
+        } while (!m_hostCriteria.compareAndSet(expect, update));
+    }
+
+    @Override
+    public JSONObject ornate(JSONObject jo, Optional<Boolean> paused) {
```
Comment (jhugg) [inline src/frontend/org/voltdb/probe/MeshProber.java]: Can we pick a more common word or several words?

ornate -> addFullVoltDBContextToJSON or something

File: src/frontend/org/voltdb/probe/HostCriteria.java (hunk 16 above)
Comment (vtkstef) [inline src/frontend/org/voltdb/probe/HostCriteria.java]: anytime we compare host criteria (testing and not)

File: src/frontend/org/voltdb/probe/MeshProber.java [hunk 26]
```diff
@@ -0,0 +1,866 @@
...
+                .put(hostId, hc)
+                .build();
+
+        } while (!m_hostCriteria.compareAndSet(expect, update));
+
+        determineStartActionIfNecessary(update);
+    }
+
+    @Override
+    public void accrue(Map<Integer, JSONObject> jos) {
```
Comment (jhugg) [inline src/frontend/org/voltdb/probe/MeshProber.java]: Is there a performance reason accrue with a map can't just call the other accrue a lot?

File: src/frontend/org/voltdb/probe/MeshProber.java (hunk 17 above)
Comment (vtkstef) [inline src/frontend/org/voltdb/probe/MeshProber.java]: anytime we compare host criteria (testing and not)

File: src/frontend/org/voltdb/probe/HostCriteria.java (hunk 21 above)
Comment (vtkstef) [inline src/frontend/org/voltdb/probe/HostCriteria.java]: it is generated by eclipse and it does follow good and widely accepted implementation guidelines

File: src/frontend/org/voltdb/probe/MeshProber.java (hunk 22 above)
Comment (vtkstef) [inline src/frontend/org/voltdb/probe/MeshProber.java]: it is generated by eclipse and it does follow good and widely accepted implementation guidelines

File: src/frontend/org/voltdb/probe/MeshProber.java (hunk 23 above)
Comment (vtkstef) [inline src/frontend/org/voltdb/probe/MeshProber.java]: it is generated by eclipse and it does follow good and widely accepted implementation guidelines

File: src/frontend/org/voltdb/probe/HostCriteria.java (hunk 21 above)
Comment (jhugg) [inline src/frontend/org/voltdb/probe/HostCriteria.java]: It doesn't follow our coding guidelines. Ifs without braces are not widely accepted.

Comment (jhugg) [inline src/frontend/org/voltdb/probe/HostCriteria.java]: I have less of an issue where the body is on the same line for stuff like this, but multi-line ifs without braces have led to bugs in our code.

File: src/frontend/org/voltdb/probe/HostCriteria.java (hunk 16 above)
Comment (jhugg) [inline src/frontend/org/voltdb/probe/HostCriteria.java]: Wouldn't that use equals?

File: src/frontend/org/voltdb/utils/CatalogUtil.java (hunk 5 above)
Comment (jhugg) [inline src/frontend/org/voltdb/utils/CatalogUtil.java]: Is the old code still around?

File: src/frontend/org/voltcore/messaging/HostMessenger.java [hunk 27]
```diff
@@ -138,8 +135,8 @@ public Config(String coordIp, int coordPort) {
         }
 
         public Config() {
-            this(null, DEFAULT_INTERNAL_PORT);
-            acceptor = MeshProber.builder()
+            this(null, org.voltdb.VoltDB.DEFAULT_INTERNAL_PORT);
```
Comment (jhugg) [inline src/frontend/org/voltcore/messaging/HostMessenger.java]: This doesn't actually fix anything. It just makes voltdb leaking into voltcore harder to notice.

File: src/frontend/org/voltdb/NodeStateTracker.java [hunk 28]
```diff
@@ -22,11 +22,14 @@
 
 import com.google_voltpatches.common.base.Supplier;
 
-public class StatusTracker {
+/**
+ * Class that aides in the tracking of a VoltDB node state.
```
Comment (jhugg) [inline src/frontend/org/voltdb/NodeStateTracker.java]: I asked for a comment that explains why this class exists, where it's used and what it does.

This comment is next to useless.

File: src/frontend/org/voltdb/common/NodeState.java [hunk 29]
```diff
@@ -24,7 +24,9 @@
 
 import com.google_voltpatches.common.collect.ImmutableList;
 
-//Enup representing state of the node
+/**
+ * Enum representing state of the node
```
Comment (jhugg) [inline src/frontend/org/voltdb/common/NodeState.java]: This kind of comment is not helpful. see previous comment.

File: src/frontend/org/voltdb/utils/CatalogUtil.java (hunk 5 above)
Comment (vtkstef) [inline src/frontend/org/voltdb/utils/CatalogUtil.java]: the one that computes a byte array hash for deployment? Yes it is used in the catalog update pathways

File: src/frontend/org/voltdb/probe/HostCriteria.java (hunk 16 above)
Comment (vtkstef) [inline src/frontend/org/voltdb/probe/HostCriteria.java]: for hash sets or collections that rely on hashCode hashCode is always called first. Here is a link with some [insights](http://stackoverflow.com/questions/27581/what-issues-should-be-considered-when-overriding-equals-and-hashcode-in-java)

Comment (jhugg) [inline src/frontend/org/voltdb/probe/HostCriteria.java]: Yes. Don't use those unless there's a good reason.

Here's the issue. Somebody adds a field in 6 months and doesn't update the hash function. BUGS BE HERE.
I ran into a bug like this a few months back and they're insidious.

File: src/frontend/org/voltdb/probe/MeshProber.java [hunk 30]
```diff
@@ -0,0 +1,879 @@
...
+        Splitter commaSplitter = Splitter.on(',').omitEmptyStrings().trimResults();
+        ImmutableSortedSet.Builder<String> sbld = ImmutableSortedSet.naturalOrder();
+        for (String h: commaSplitter.split(option)) {
+            checkArgument(isValidCoordinatorSpec(h), "%s is not a valid host spec", h);
+            sbld.add(HostAndPort.fromString(h).withDefaultPort(DEFAULT_INTERNAL_PORT).toString());
+        }
+        return sbld.build();
+    }
+
+    public static ImmutableSortedSet<String> hosts(int...ports) {
```
Comment (nshi) [inline src/frontend/org/voltdb/probe/MeshProber.java]: What's this used for without passing any IP/hostname?

File: src/frontend/org/voltcore/messaging/JoinAcceptor.java [hunk 31]
```diff
@@ -0,0 +1,163 @@
...
+     */
+    default NavigableSet<String> getCoordinators() {
+        return ImmutableSortedSet.of("");
+    }
+
+    /**
+     * notifies the acceptor when a host is removed from the mesh
+     * @param hostId
+     */
+    public void detract(int hostId);
```
Comment (nshi) [inline src/frontend/org/voltcore/messaging/JoinAcceptor.java]: Kind of a weird name. It's not obvious to me that it's just removing the host.

File: src/frontend/org/voltcore/messaging/JoinAcceptor.java [hunk 32]
```diff
@@ -0,0 +1,163 @@
...
+ */
+public interface JoinAcceptor extends JSONString {
+
+    /**
+     * Give the acceptor the {@link JSONObject} provided at the initial connection handshake
+     *
+     * @param hostId host id for the new connection
+     * @param jo the {@link JSONObject} provided at the initial connection handshake
+     */
+    public void accrue(int hostId, JSONObject jo);
```
Comment (nshi) [inline src/frontend/org/voltcore/messaging/JoinAcceptor.java]: Is this just adding the host to the acceptor's known hosts?

File: src/frontend/org/voltdb/probe/MeshProber.java [hunk 33]
```diff
@@ -0,0 +1,879 @@
...
+        Map<Integer,HostCriteria> expect;
+        Map<Integer,HostCriteria> update;
+
+        do {
+            expect = m_hostCriteria.get();
+            update = ImmutableMap.<Integer, HostCriteria>builder()
+                .putAll(Maps.filterKeys(expect, not(in(hostIds))))
+                .build();
+        } while (!m_hostCriteria.compareAndSet(expect, update));
+    }
```
Comment (nshi) [inline src/frontend/org/voltdb/probe/MeshProber.java]: The two implementations of `detract()` look largely the same. Can this one call the other?

File: src/frontend/org/voltdb/probe/MeshProber.java [hunk 34]
```diff
@@ -0,0 +1,879 @@
...
+            expect = m_hostCriteria.get();
+            update = ImmutableMap.<Integer, HostCriteria>builder()
+                .putAll(Maps.filterKeys(expect, not(in(additions.keySet()))))
+                .putAll(additions)
+                .build();
+
+        } while (!m_hostCriteria.compareAndSet(expect, update));
+
+        determineStartActionIfNecessary(update);
+    }
```
Comment (nshi) [inline src/frontend/org/voltdb/probe/MeshProber.java]: Same here, almost the same code as the other implementation.

File: src/frontend/org/voltdb/probe/MeshProber.java [hunk 35]
```diff
@@ -0,0 +1,879 @@
...
+            if (c.getNodeState().operational()) {
+                operational += 1;
+                // pause state from operational nodes overrides yours
+                paused = operational == 1 ? c.isPaused() : paused;
+            }
+            unmeshed += c.getNodeState().unmeshed() ? 1 : 0;
+            bare += c.isBare() ? 1 : 0;
+            if (c.isPaused() && operational == 0) {
+                paused = c.isPaused();
+            }
```
Comment (nshi) [inline src/frontend/org/voltdb/probe/MeshProber.java]: You can move this if block to the top of the for loop and drop line 512.

By the way, this seem to depend on which host is the first in the unsorted map, which is nondeterministic.

File: src/frontend/org/voltdb/probe/MeshProber.java [hunk 36]
```diff
@@ -0,0 +1,879 @@
...
+
+    public static MeshProber prober(HostMessenger hm) {
+        return MeshProber.class.cast(hm.getAcceptor());
+    }
+
+    protected final ImmutableSortedSet<String> m_coordinators;
+    protected final VersionChecker m_versionChecker;
+    protected final boolean m_enterprise;
+    protected final StartAction m_startAction;
+    protected final boolean m_bare;
```
Comment (nshi) [inline src/frontend/org/voltdb/probe/MeshProber.java]: What does bare mean?

File: src/frontend/org/voltdb/probe/MeshProber.java [hunk 37]
```diff
@@ -0,0 +1,879 @@
...
+
+        StartAction determination = isBare() ?
+                  StartAction.CREATE
+                : StartAction.RECOVER;
+
+        if (operational > 0 && operational < hostCount) { // rejoin
+            determination = StartAction.LIVE_REJOIN;
+        } else if (operational > 0 && operational == hostCount) { // join
+            if (isAddAllowed()) {
+                hostCount = hostCount + ksafety;
```
Comment (nshi) [inline src/frontend/org/voltdb/probe/MeshProber.java]: Shouldn't it be `hostCount = hostCount + ksafety + 1;`?

File: src/frontend/org/voltdb/probe/MeshProber.java [hunk 38]
```diff
@@ -0,0 +1,879 @@
...
+                determination = StartAction.JOIN;
+            } else {
+                org.voltdb.VoltDB.crashLocalVoltDB("Node is not allowed to rejoin an already whole cluster");
```
Comment (nshi) [inline src/frontend/org/voltdb/probe/MeshProber.java]: Inconsistent error message.

File: src/frontend/org/voltdb/RealVoltDB.java [hunk 39]
```diff
@@ -154,22 +165,22 @@
  * namespace. A lot of the global namespace is described by VoltDBInterface
  * to allow test mocking.
  */
-public class RealVoltDB implements VoltDBInterface, RestoreAgent.Callback, HostMessenger.MembershipAcceptor, HostMessenger.HostWatcher {
-    private static final boolean DISABLE_JMX = Boolean.valueOf(System.getProperty("DISABLE_JMX", "false"));
+public class RealVoltDB implements VoltDBInterface, RestoreAgent.Callback, HostMessenger.HostWatcher {
+    private static final boolean DISABLE_JMX = Boolean.valueOf(System.getProperty("DISABLE_JMX", "true"));
```
Comment (nshi) [inline src/frontend/org/voltdb/RealVoltDB.java]: Is this something we have to release note? Customers may be using it.

File: src/frontend/org/voltdb/probe/MeshProber.java (hunk 30 above)
Comment (vtkstef) [inline src/frontend/org/voltdb/probe/MeshProber.java]: I'll comment that is mainly a test artifact (LocalCluster)

File: src/frontend/org/voltdb/RealVoltDB.java [hunk 40]
```diff
@@ -502,8 +563,33 @@ public void initialize(VoltDB.Configuration config) {
...
             }
 
-            buildClusterMesh(isRejoin || m_joining);
+            MeshProber.Determination determination = buildClusterMesh(readDepl);
+            if (m_config.m_startAction == StartAction.PROBE) {
+                String action = "Starting a new database cluster";
+                if (determination.startAction.doesRejoin()) {
+                    action = "Rejoining a running cluster";
+                } else if (determination.startAction.doesRecover()) {
+                    action = "Restating the database cluster from the command logs";
```
Comment (nshi) [inline src/frontend/org/voltdb/RealVoltDB.java]: Typo, should be "Restarting"

**Comment by vtkstef** [comment]
--squash merged
